unit_tests/sources/declarative/incremental/test_concurrent_perpartitioncursor.py

```diff
@@ -3773,3 +3788,196 @@ def test_given_global_state_when_read_then_state_is_not_per_partition() -> None:
+    )
+
+    # ── Partition sequence: 1 → 2 → 1 ──────────────────────────────────
+    partitions = [
+        StreamSlice(partition={"id": "1"}, cursor_slice={}, extra_fields={}),
+        StreamSlice(partition={"id": "2"}, cursor_slice={}, extra_fields={}),
+        StreamSlice(partition={"id": "1"}, cursor_slice={}, extra_fields={}),
+    ]
+    pr = cursor._partition_router
+    pr.stream_slices.return_value = iter(partitions)
+    pr.get_stream_state.return_value = {}
+
+    # Iterate lazily so that the first "1" gets cleaned before
+    # the second "1" arrives.
+    slice_gen = cursor.stream_slices()
+
+    first_1 = next(slice_gen)
+    cursor.close_partition(
+        DeclarativePartition("dup_stream", {}, MagicMock(), MagicMock(), first_1)
+    )
+
+    two = next(slice_gen)
+    cursor.close_partition(DeclarativePartition("dup_stream", {}, MagicMock(), MagicMock(), two))
+
+    # Second “1” should appear because the semaphore was cleaned up
+    second_1 = next(slice_gen)
+    cursor.close_partition(
+        DeclarativePartition("dup_stream", {}, MagicMock(), MagicMock(), second_1)
+    )
+
+    with pytest.raises(StopIteration):
+        next(slice_gen)
+
+    assert cursor._IS_PARTITION_DUPLICATION_LOGGED is False  # no duplicate warning
+    assert len(cursor._cursor_per_partition) == 2  # only “1” & “2” kept
+    assert len(cursor._semaphore_per_partition) == 0  # all semaphores cleaned
+    assert len(cursor._processing_partitions_indexes) == 0
+    assert len(cursor._partition_key_to_index) == 0
+    assert len(cursor._partitions_done_generating_stream_slices) == 0
+
+
+def test_duplicate_partition_while_processing():
+    inner_cursors = [
+        _make_inner_cursor("2024-01-01T00:00:00Z"),  # first “1”
+        _make_inner_cursor("2024-01-02T00:00:00Z"),  # “2”
+        _make_inner_cursor("2024-01-03T00:00:00Z"),  # for second "1"
+    ]
+
+    factory = MagicMock()
+    factory.create.side_effect = inner_cursors
+
+    cursor = ConcurrentPerPartitionCursor(
+        cursor_factory=factory,
+        partition_router=MagicMock(),
+        stream_name="dup_stream",
+        stream_namespace=None,
+        stream_state={},
+        message_repository=MagicMock(),
+        connector_state_manager=MagicMock(),
+        connector_state_converter=MagicMock(),
+        cursor_field=CursorField(cursor_field_key="updated_at"),
+    )
+
+    partitions = [
+        StreamSlice(partition={"id": "1"}, cursor_slice={}, extra_fields={}),
+        StreamSlice(partition={"id": "2"}, cursor_slice={}, extra_fields={}),
+        StreamSlice(partition={"id": "1"}, cursor_slice={}, extra_fields={}),
+    ]
+    pr = cursor._partition_router
+    pr.stream_slices.return_value = iter(partitions)
+    pr.get_stream_state.return_value = {}
+
+    generated = list(cursor.stream_slices())
+    # Only “1” and “2” emitted – duplicate “1” skipped
+    assert len(generated) == 2
+
+    # Close “2” first
+    cursor.close_partition(
```

Comment:
After closing 2, do we clean the semaphore for 2 as well? I would assume not because 1 hasn't been completed on the second parent record `1` but I fear this might lead to partitions not being cleaned in some very specific case. For example, if even other parent record is the first parent record like [1, 2, 1, 3, 1, 4, 1, 5, 1, 6, 1]. In this case, since 1 pops up frequently, we won't be able to clean 2, 3, 4, 5 and 6.